clustering/web/WFLY-5497_Distributable_Shared_Session_Config.adoc

```diff
@@ -0,0 +1,59 @@
+= Distributable Shared Session Configuration
+:author:            Paul Ferraro
+:email:             paul.ferraro@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+Currently, the shared-session-config schema (used to enable use of session sharing for WARs within an EAR) does not indicate whether a distributable or non-distributable session manager should be used.
+Consequently, we assign the session manager implementation based on the availability of the module providing the distributed implementation.
+Since the default infinispan web session configuration attempts to passivate sessions on shutdown, users will see NotSerializableExceptions on shutdown if they use non-serializable session attributes - even if their web applications do not define themselves as being <distributable/>.
+The only existing workaround for this is to remove the requisite clustering modules from the server - which of course means that the same server cannot support any traditional distributable web applications.
+To fix this, we should add a <distributable/> element (a la web.xml) to the shared-session-config schema to determine which session manager implementation to use.
+
+To retain compatibility, version 1.0 of the schema will continue presume to use a distributable shared session manager.
+Beginning with version 2.0, users will need to opt-in to this behavior via the new <distributable/> element (as we already do for individual web applications).
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFLY-5497
+
+=== Related Issues
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:tborgato@redhat.com[Tommasso Borgato]
+
+=== Affected Projects or Components
+
+* Clustering, Undertow subsystem
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+* Add new version 2.0 of shared-session-config schema, including new <distributable/> element.
+* Add boolean isDistributable() method to SharedSessionConfig meta-data.
+* isDistributable() should return true if version 1.0 of the schema was specified.
+* SharedSessionConfig DUP should only install a distributable session manager if isDistributable() returns true.
+
+=== Nice-to-Have Requirements
+
+* Parameterize existing shared-session-config parser by schema version.
+
+=== Non-Requirements
+
+== Test Plan
+
+* Modify existing shared session config clustering test to use new schema version.
```

Comment:
In general, the integration testsuite only covers the most recent version of deployment descriptors.  It is not feasible to test every combination of deployment descriptor versions within the context of an integration testsuite, otherwise test runs would take several orders of magnitude longer to run.  The parsing of old versions of deployment descriptors are validated by unit tests.

Reply:
Usually I would agree, but in this case new schema version changes default behavior. I don't think it is desirable to lost test coverage in upstream test suite.

Reply:
There is no loss in test coverage.  The unit test verifies the parsing output for both schema versions.
See: https://github.com/pferraro/wildfly/blob/shared/undertow/src/test/java/org/wildfly/extension/undertow/session/SharedSessionConfigXMLReaderTestCase.java